components/org.wso2.carbon.uis/src/main/java/org/wso2/carbon/uis/internal/io/deployment/ArtifactAppDeployer.java

```diff
@@ -27,125 +27,170 @@
-        for (String appName : deployedAppNames) {
-            publishAppUndeploymentEvent(appName);
+    @Override
+    public void undeploy(Object key) throws CarbonDeploymentException {
+        App app = deployedApps.remove(key.toString());
+        if (app == null) {
+            throw new CarbonDeploymentException("Web app for key '{}' cannot be found to undeploy.");
         }
+
+        appDeploymentEventListener.appUndeploymentEvent(app.getName());
+        LOGGER.debug("Web app '{}' undeployed from context path '{}'.", app.getName(), app.getContextPath());
     }
 
-    private void publishAppDeploymentEvent(AppReference deployingAppReference) {
-        for (AppDeploymentEventListener appDeploymentEventListener : eventListeners) {
-            appDeploymentEventListener.appDeploymentEvent(deployingAppReference);
-        }
-        deployedAppNames.add(deployingAppReference.getName());
+    @Override
+    public Object update(Artifact artifact) throws CarbonDeploymentException {
+        // when an app artifact gets updated, we just re-deploy it
+        undeploy(artifact.getKey());
+        return deploy(artifact);
     }
 
-    private void publishAppsDeploymentEvents(Set<AppReference> deployingAppsReferences) {
-        eventListeners.forEach(listener -> listener.appsDeploymentEvents(deployingAppsReferences));
+    @Override
+    public URL getLocation() {
+        return deploymentLocation;
     }
 
-    private void publishAppUndeploymentEvent(String undeployingAppName) {
-        for (AppDeploymentEventListener appDeploymentEventListener : eventListeners) {
-            appDeploymentEventListener.appUndeploymentEvent(undeployingAppName);
+    @Override
+    public ArtifactType getArtifactType() {
+        return artifactType;
+    }
+
+    private boolean isValidAppArtifact(Path appPath) throws CarbonDeploymentException {
+        try {
+            return Files.exists(appPath) && Files.isDirectory(appPath) && Files.isReadable(appPath) &&
+                   !Files.isHidden(appPath);
+        } catch (IOException e) {
+            throw new CarbonDeploymentException("Cannot access web app artifact in '" + appPath + "'.", e);
         }
-        deployedAppNames.remove(undeployingAppName);
+    }
+
+    private App createApp(Path appPath) throws CarbonDeploymentException {
+        AppReference appReference = new ArtifactAppReference(appPath);
+        String appContextPath = createAppContextPath(appReference);
+        try {
+            return AppCreator.createApp(appReference, appContextPath);
+        } catch (AppCreationException e) {
+            throw new CarbonDeploymentException(
+                    "Cannot create web app '" + appReference.getName() + "' from artifact '" + appReference.getPath() +
+                    "' to deploy for context path '" + appContextPath + "'.", e);
+        }
+    }
+
+    private String createAppContextPath(AppReference appReference) {
+        return "/" + appReference.getName();
```

Comment:
Why "/" ? Does this work in Windows?

Reply:
Why do u need a helper method here?

Reply:
>Why "/" ? Does this work in Windows?

App context path is part of the URL for a page, not a file path. e.g. https://localhost:9090/test/home here `/test` is the app context path. 

>Why do u need a helper method here?

This method is responsible for returning the app context path for a given app reference. Currently app context path is simply `/<app_name>` which is the default value. But we are hoping to give the capability to override the app context path via the `deployment.yaml`. So in future, this method will have the logic to return the overridden value from `deployment.yaml` or to return the default value.

Reply:
Added a TODO in commit 8d928e8e34282d7ad528d448d15f030347952813